pymc_marketing/mmm/delayed_saturated_mmm.py

```diff
@@ -63,80 +63,147 @@ def __init__(
-        else:
-            control_data = None
+            control_data = X[self.control_columns]
+            coords["control"] = self.control_columns
+            X_data = pd.concat([X_data, control_data], axis=1)
 
+        fourier_features: Optional[pd.DataFrame] = None
         if self.yearly_seasonality is not None:
-            fourier_features = self._get_fourier_models_data()
+            fourier_features = self._get_fourier_models_data(X=X)
+            self.fourier_columns = fourier_features.columns
             coords["fourier_mode"] = fourier_features.columns.to_numpy()
+            X_data = pd.concat([X_data, fourier_features], axis=1)
 
-        else:
-            fourier_features = None
+        self.model_coords = coords
+        if self.validate_data:
+            self.validate("X", X_data)
+            self.validate("y", y)
+        self.preprocessed_data["X"] = self.preprocess("X", X_data.copy())
+        self.preprocessed_data["y"] = self.preprocess("y", y.copy())
+        self.X: pd.DataFrame = X_data
+        self.y: pd.Series = y
+
+    def _preprocess_channel_prior(self) -> TensorVariable:
+        return (
+            pm.HalfNormal.dist(sigma=2, shape=len(self.channel_columns))
+            if self.channel_prior is None
+            else change_dist_size(
+                dist=self.channel_prior, new_size=len(self.channel_columns)
+            )
+        )
 
-        with pm.Model(coords=coords) as self.model:
+    def build_model(
+        self,
+        X: pd.DataFrame,
+        y: pd.Series,
+        **kwargs,
+    ) -> None:
+        model_config = self.model_config
+        self.generate_and_preprocess_model_data(X, y)
+        with pm.Model(coords=self.model_coords) as self.model:
             channel_data_ = pm.MutableData(
                 name="channel_data",
-                value=channel_data,
+                value=self.X[self.channel_columns].to_numpy(),
                 dims=("date", "channel"),
             )
 
-            target_ = pm.MutableData(name="target", value=target_data, dims="date")
+            target_ = pm.MutableData(
+                name="target",
+                value=self.y,
+                dims="date",
+            )
 
-            intercept = pm.Normal(name="intercept", mu=0, sigma=2)
+            intercept = pm.Normal(
+                name="intercept",
+                mu=model_config["intercept"]["mu"],
+                sigma=model_config["intercept"]["sigma"],
+            )
 
+            """beta_channel = pm.HalfNormal(
+                name="beta_channel",
+                sigma=model_config["beta_channel"]["sigma"],
+                dims=model_config["beta_channel"]["dims"],
+            )  # ? Allow prior depend on channel costs?"""
             channel_prior = self._preprocess_channel_prior()
             beta_channel = self.model.register_rv(
                 rv_var=channel_prior, name="beta_channel", dims="channel"
```

Comment:
@juanitorduz I need your input on this one, as the change was introduced by you. "beta)channel" when uncommented throws an error that the name is already registered. How should I modify it so it works as intended and doesn't crash?

Reply:
The whole point of lines 180 - 182 is to add custom priors (see unit test and notebook example). As far as I understand the model builder approach already supports custom priors right? In this case, you can simply delete the lines added in  https://github.com/pymc-labs/pymc-marketing/pull/286 (see also feedback from https://github.com/pymc-labs/pymc-marketing/issues/294)

Reply:
ok, beta_priors were removed and the build_model lines creating beta_channel based on model_config were uncommented